### PR TITLE
Add cloudevents-spring sample

### PR DESCRIPTION
**Pull request by dsyer**
Same basic features as the other samples. Includes README
with details. CloudEvents support in Spring Cloud Function
is not GA, but the sample depends on a stable milestone.
I can send another PR when the GA release comes.

### CODE REVIEW

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+1. The default mode has the app reply to your input events with the output
+   event, which is simplest for demonstrating things working in isolation, but
+   is also the model for working for the Knative Eventing `Broker` concept. The
+   input event is modified assigning a new source and type attribute.
+
+2. `K_SINK` mode has the app send events to the destination encoded in
+   `$K_SINK`, which is useful to demonstrate how folks can synthesize events to
+   send to a Service or Broker when not initiated by a Broker invocation (e.g.
+   implementing an event source). The input event is modified assigning a new
```
Comment (vaikas) [inline docs/serving/samples/cloudevents/cloudevents-spring/README.md:14]: In this mode, there's no input event? left over?

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+   source and type attribute.
+
+The application will use `$K_SINK`-mode whenever the environment variable is
```
Comment (vaikas) [inline docs/serving/samples/cloudevents/cloudevents-spring/README.md:17]: nit for consistency, maybe drop the $ since above you say K_SINK mode?

**Review by vaikas — COMMENTED** [review]
Thanks!

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md (hunk 1 above)
Comment (dsyer) [reply docs/serving/samples/cloudevents/cloudevents-spring/README.md:14]: I copied the README from another sample. So you're asking the wrong person, but there is an input event, in the sense that the application accepts input. It seems to work teh same as the other sample when I run it locally, so I might be wrong, but then we both are.

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md (hunk 2 above)
Comment (dsyer) [reply docs/serving/samples/cloudevents/cloudevents-spring/README.md:17]: Copy paste. If you want to fix this sample, you would want to fix the other(s). Can we make that a separate patch?

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md (hunk 1 above)
Comment (vaikas) [reply docs/serving/samples/cloudevents/cloudevents-spring/README.md:14]: I was confused, I thought this was acting as a source, in that there's no incoming event. You're right, sorry for the confusion :)

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
+# Spring + CloudEvents + Knative
+
+A simple web app written in Java using Spring Cloud Function that can receive CloudEvents. It
+supports running in two modes:
+
+1. The default mode has the app reply to your input events with the output
+   event, which is simplest for demonstrating things working in isolation, but
+   is also the model for working for the Knative Eventing `Broker` concept. The
```
Comment (mpetason) [inline docs/serving/samples/cloudevents/cloudevents-spring/README.md:8]:
```suggestion
   is also the model for working with the Knative Eventing `Broker` concept. The
```

Not sure if it should be "the model of working for the Knative Eventing Broker" - for working for sounds a little bit off, but I'm not sure what would make it clear. 

Maybe even "but is also the model for the Knative Evenging `Broker` concept. Could drop "for working"

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+## Build and deploy the sample
+
+To build the image, run:
+
+```shell
+mvn compile jib:build -Dimage=<image_name>
+```
+
+To deploy the Knative Service, look in the `service.yaml` and replace `<image>`
```
Comment (mpetason) [inline docs/serving/samples/cloudevents/cloudevents-spring/README.md:45]:
```suggestion
To deploy the Knative Service, edit the `service.yaml` file and replace `<image>`
```

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+      "specversion": "1.0",
+      "source": "http://curl-command",
+      "type": "curl.demo",
+      "id": "123-abc",
+      "data": {"name":"Dave"}
+    }' \
+    http://cloudevents-java.xip.io
+```
+
+N.B. the response from the server is "binary", not "structured", so you get the same response as if you sent a binary message..
```
Comment (mpetason) [inline docs/serving/samples/cloudevents/cloudevents-spring/README.md:90]: Maybe expand what N.B. stands for, I'm not sure of what it means in the sentence.

File: docs/serving/samples/cloudevents/cloudevents-spring/index.md [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+weight: 1
+type: "docs"
+---
+
+A simple web app written in Java using Spring Cloud Function that can receive CloudEvents. It
+supports running in two modes:
+
+1. The default mode has the app reply to your input events with the output
+   event, which is simplest for demonstrating things working in isolation, but
+   is also the model for working for the Knative Eventing `Broker` concept. The
```
Comment (mpetason) [inline docs/serving/samples/cloudevents/cloudevents-spring/index.md:13]:
```suggestion
   is also the model for working with the Knative Eventing `Broker` concept. The
```

^ whichever update is used above can be used here.

File: docs/serving/samples/cloudevents/cloudevents-spring/index.md [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
+
+## Build and deploy the sample
+
+To build the image, run:
+
+```shell
+mvn compile jib:build -Dimage=<image_name>
+```
+
+To deploy the Knative Service, look in the `service.yaml` and replace `<image>`
```
Comment (mpetason) [inline docs/serving/samples/cloudevents/cloudevents-spring/index.md:50]:
```suggestion
To deploy the Knative Service, edit the `service.yaml` file and replace `<image>`
```

**Review by mpetason — COMMENTED** [review]
Looks great. Added a few edits.

File: docs/serving/samples/cloudevents/cloudevents-spring/README.md (hunk 5 above)
Comment (dsyer) [reply docs/serving/samples/cloudevents/cloudevents-spring/README.md:90]: N.B. -> "Note". I'll put in another commit and then squash them all down.

**Comment by dsyer** [comment]
Note that some of the README comments above apply to the original document (the cloudevents-vertx sample).

**Comment by dsyer** [comment]
Latest commit adds integration test for K_SINK use case.

**Comment by mattmoor** [comment]
/assign @vaikas @mpetason 

Any final comments from folks?

**Comment by mpetason** [comment]
Nothing from me. Looks like the updates I requested were resolved. I'll look into updating the docs this pulls from since some of the fixes were just updating the original template.

**Pull request merged**